format/bplist/bplist.go

```diff
@@ -0,0 +1,245 @@
+	elementTypeArray            = 0x0A
+	elementTypeSet              = 0x0C
+	elementTypeDict             = 0x0D
+)
+
+const (
+	null      = 0x00
+	boolFalse = 0x08
+	boolTrue  = 0x09
+)
+
+var singleByteElementMap = scalar.UToScalar{
+	null:      {Sym: "null", Description: "Null Value"},
+	boolFalse: {Sym: "false", Description: "False Value"},
+	boolTrue:  {Sym: "true", Description: "True Value"},
+}
+
+var elementTypeMap = scalar.UToScalar{
+	elementTypeNullOrBoolOrFill: {Sym: "singleton", Description: "Singleton value (null/bool)"},
+	elementTypeInt:              {Sym: "int", Description: "Integer"},
+	elementTypeReal:             {Sym: "real", Description: "Floating Point Number"},
+	elementTypeDate:             {Sym: "date", Description: "Date, 8-byte float"},
+	elementTypeData:             {Sym: "data", Description: "Binary data"},
+	elementTypeASCIIString:      {Sym: "ascii_string", Description: "ASCII encoded string"},
+	elementTypeUnicodeString:    {Sym: "unicode_string", Description: "Unicode string"},
+	elementTypeUID:              {Sym: "uid", Description: "UID"},
+	elementTypeArray:            {Sym: "array", Description: "Array"},
+	elementTypeSet:              {Sym: "set", Description: "Set"},
+	elementTypeDict:             {Sym: "dict", Description: "Dictionary"},
+}
+
+// decodes the number of bits required to store the following object
+func decodeSize(d *decode.D) uint64 {
+	n := d.FieldU4("size_bits")
+	if n != 0x0F {
+		return uint64(n)
+	}
+
+	d.FieldU4("large_size_marker", d.AssertU(1)) // TODO: add assertion that this is 0001
+
+	// get the exponent value
+	n = d.FieldU4("exponent")
+
+	// calculate the number of bytes encoding the size
+	n = uint64(math.Pow(2, float64(n)))
+
+	// decode that many bytes as big endian
+	n = d.FieldUBigIntBE("size", int(n*8)).Uint64()
+	return n
+}
+
+func decodeItem(d *decode.D, p *plist) any {
+	d.FieldStruct("object", func(d *decode.D) {
+		m := d.FieldU4("type", elementTypeMap)
+		switch m {
+		case elementTypeNullOrBoolOrFill:
+			t := d.U4()
+			switch t {
+			case null:
+				d.FieldValueNil("value")
+			case boolTrue:
+				d.FieldValueBool("value", true)
+			case boolFalse:
+				d.FieldValueBool("value", false)
+			}
+		case elementTypeInt:
+			n := decodeSize(d)
+			d.FieldSFn("value", func(d *decode.D) int64 {
+				return d.UBigIntBE(int(n)).Int64()
+			})
+		case elementTypeReal:
+			n := decodeSize(d)
+			d.FieldFFn("value", func(d *decode.D) float64 {
+				return d.FE(int(n), decode.BigEndian)
+			})
+		case elementTypeDate:
+			d.FieldStrFn("value", func(d *decode.D) string {
+				v := d.F64()
+				t := time.Unix(int64(v), 0)
+				return t.String()
+			})
+		case elementTypeData:
+			d.FieldStruct("value", func(d *decode.D) {
+				n := decodeSize(d)
+				d.FieldRawLen("value", int64(n))
+			})
+		case elementTypeASCIIString:
+			d.FieldStruct("value", func(d *decode.D) {
+				d.FieldStrFn("value", func(d *decode.D) string {
+					n := decodeSize(d)
```

Comment:
I noticed that the range for string gets a bit weird with this, it looks like the string start 4 bits earlier then it actually do. Maybe better to move out decodeSize? some like:
```go
d := decodeSize(d)
d.FieldUTF8("value", int(n));
```
or if decode size is not decoded in parts (see earlier comments):
```go
d := d.FieldUFn("size", decodeSize(d))
d.FieldUTF8("value", int(n));
```
but then `decodeSize` needs to make sure to read the 4 "padding" bits if size_bits is < 0xf to make offset correct for the string